### PR TITLE
Add common changes to support Azure Batch AAD

### PR DESCRIPTION
**Pull request by matthchr**
## Description

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [ ] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [ ] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [ ] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

**Comment by matthchr** [comment]
This PR is in preparation for adding Batch AAD support. See #4762 for the official support.

**Comment by matthchr** [comment]
@markcowl please take a look when you get a chance

File: src/ResourceManager/Profile/Commands.Profile.Test/AzureRMProfileTests.cs [hunk 1]
```diff
@@ -733,6 +733,7 @@ public void SavingProfileWorks()
       ""AzureKeyVaultServiceEndpointResourceId"": null,
       ""GraphEndpointResourceId"": null,
       ""DataLakeEndpointResourceId"": null,
+      ""BatchEndpointResourceId"": null,
```
Comment (markcowl) [inline src/ResourceManager/Profile/Commands.Profile.Test/AzureRMProfileTests.cs]: can we add in tests that verify this value - we have tests that check the output type, and tests that check the value that is saved in the context

File: src/ResourceManager/Profile/Commands.Profile/Environment/AddAzureRMEnvironment.cs [hunk 2]
```diff
@@ -146,6 +146,11 @@ public string DataLakeAudience
             }
         }
 
+        [Parameter(Position = 20, Mandatory = false, ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The resource identifier of the Azure Batch service that is the recipient of the requested token.")]
+        [Alias("BatchResourceId")]
```
Comment (markcowl) [inline src/ResourceManager/Profile/Commands.Profile/Environment/AddAzureRMEnvironment.cs]: I think we want to use BatchAudience as well

File: src/ResourceManager/Profile/Commands.Profile/Environment/AddAzureRMEnvironment.cs [hunk 3]
```diff
@@ -285,6 +290,8 @@ public override void ExecuteCmdlet()
                                    nameof(GraphAudience));
                                 SetEndpointIfBound(newEnvironment, AzureEnvironment.Endpoint.DataLakeEndpointResourceId,
                                     nameof(DataLakeAudience));
+                                SetEndpointIfBound(newEnvironment, AzureEnvironment.Endpoint.BatchEndpointResourceId,
```
Comment (markcowl) [inline src/ResourceManager/Profile/Commands.Profile/Environment/AddAzureRMEnvironment.cs]: Note, I think you will also need to change the constructors and type conversion operators for PSAzureEnvinronment to include this value.  If you add this value into existing test validation overt these types, you'll see the code you need to change.

File: src/ResourceManager/Profile/Commands.Profile/Environment/AddAzureRMEnvironment.cs (hunk 2 above)
Comment (matthchr) [reply src/ResourceManager/Profile/Commands.Profile/Environment/AddAzureRMEnvironment.cs]: As an alias you mean? Will add it.

File: src/ResourceManager/Profile/Commands.Profile.Test/AzureRMProfileTests.cs (hunk 1 above)
Comment (matthchr) [reply src/ResourceManager/Profile/Commands.Profile.Test/AzureRMProfileTests.cs]: Okay, I have updated the `TypeConversionTests` as well (I hadn't realized they were there earlier).

Can you point me to the test classes/methods for any other tests which I should update?

File: src/ResourceManager/Profile/Commands.Profile/Environment/AddAzureRMEnvironment.cs (hunk 3 above)
Comment (matthchr) [reply src/ResourceManager/Profile/Commands.Profile/Environment/AddAzureRMEnvironment.cs]: This was already done I think, right? I made the change to `PSAzureEnvironment(PSObject other)` already, and to `Equals` in `PSAzureEnvironment` - Did I need to do something else too?

File: src/ResourceManager/Profile/Commands.Profile/Environment/AddAzureRMEnvironment.cs (hunk 2 above)
Comment (matthchr) [reply src/ResourceManager/Profile/Commands.Profile/Environment/AddAzureRMEnvironment.cs]: This is done

**Comment by markcowl** [comment]
on demand run here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/73/

**Comment by matthchr** [comment]
@markcowl It looks like the on-demand run failed? I am not sure if it's related to this change or not though (logs are hard to parse...)
